### PR TITLE
assistant — fix timing issues & invalid context

### PR DESCRIPTION
**Pull request by theomonnom**
- fix invalid context with long speech 
- make sure only one speech can start at a time
- fix tts_forwarder.aclose never ending
      - also added segment_playout_finished to notify when a segment ended playing (useful for synchronizing close)
- fix _play_speech_if_validated_task never ending because playout_tx.close was never called when synthesize is cancelled
- added TaskSet to create task in a fire-and-forget fashion (you still have to aclose the TaskSet — wdyt about this the aclose fnc API?)
- fixed some misc typing stuff

### CODE REVIEW

File: livekit-agents/livekit/agents/tokenize/tokenizer.py [hunk 1]
```diff
@@ -24,13 +24,13 @@ def push_text(self, text: str | None) -> None: ...
     def mark_segment_end(self) -> None:
         self.push_text(None)
 
-    async def aclose(self, *, wait: bool = True) -> None: ...
```
Comment (theomonnom) [inline livekit-agents/livekit/agents/tokenize/tokenizer.py:27]: This was never used

**Comment by ubeytd** [comment]
Great thanks 🙏🏻! When will this be released approximately?

**Comment by theomonnom** [comment]
> Great thanks 🙏🏻! When will this be released approximately?

Hey, as soon as this is approved, it may be today when the folks in the Baie Area wake up :)

**Review by keepingitneil — APPROVED** [review]

File: livekit-agents/livekit/agents/transcription/tts_forwarder.py [hunk 2]
```diff
@@ -123,231 +115,217 @@ def __init__(
...
-            await asyncio.sleep(self._opts.new_sentence_delay)
-            seg.sentence_count += 1
+            print("segment finished")
 
-        while True:
-            audio_seg = await self._seg_queue.get()
-            if audio_seg is None:
+        rtc_seg_q.put_nowait(None)
+        await forward_task
+        print("task done")
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/transcription/tts_forwarder.py:247]: remove debug?

File: livekit-agents/livekit/agents/transcription/tts_forwarder.py [hunk 3]
```diff
@@ -123,231 +115,217 @@ def __init__(
...
+            sentence_stream = seg.sentence_stream
+            seg.forward_start_time = time.time()
+
+            async for ev in sentence_stream:
+                if ev.type == tokenize.TokenEventType.TOKEN:
+                    await self._sync_sentence_co(seg, ev.token, rtc_seg_q)
 
-            await asyncio.sleep(self._opts.new_sentence_delay)
-            seg.sentence_count += 1
+            print("segment finished")
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/transcription/tts_forwarder.py:243]: remove debug?

File: livekit-agents/livekit/agents/voice_assistant/assistant.py [hunk 4]
```diff
@@ -532,15 +532,15 @@ def _on_final_transcript(self, text: str) -> None:
...
+            user_question=self._transcripted_text,
         )
 
         # this speech may not be validated, so we create a copy
         # of our context to add the new user message
         copied_ctx = self._chat_ctx.copy()
         copied_ctx.messages.append(
             allm.ChatMessage(
-                text=text,
+                text=self._transcripted_text,
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/voice_assistant/assistant.py:543]: nit, spelling

```suggestion
                text=self._transcribed_text,
```

**Review by davidzhao — APPROVED** [review]
lgtm! just a couple of nits

**Pull request merged**